pgmpy/factors/distributions/CustomDistribution.py

```diff
@@ -0,0 +1,509 @@
+
+        *args, **kwargs: values
+            The parameters to be given to the Discretizer Class.
+
+        Returns
+        -------
+        An n-D array or a DiscreteFactor object according to the discretiztion
+        method used.
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> from scipy.special import beta
+        >>> from pgmpy.factors.continuous import ContinuousFactor
+        >>> from pgmpy.factors.continuous import RoundingDiscretizer
+        >>> def dirichlet_pdf(x, y):
+        ...     return (np.power(x, 1) * np.power(y, 2)) / beta(x, y)
+        >>> dirichlet_factor = ContinuousFactor(['x', 'y'], dirichlet_pdf)
+        >>> dirichlet_factor.discretize(RoundingDiscretizer,
+        ...                             low=1, high=2, cardinality=5)
+        # TODO: finish this
+        """
+        super(CustomDistribution, self).discretize(method, *args, **kwargs)
+
+    def reduce(self, values, inplace=True):
+        """
+        Reduces the factor to the context of the given variable values.
+
+        Parameters
+        ----------
+        values: list, array-like
+            A list of tuples of the form (variable_name, variable_value).
+
+        inplace: boolean
+            If inplace=True it will modify the factor itself, else would return
+            a new CustomDistribution object.
+
+        Returns
+        -------
+        CustomDistribution or None:
+                    if inplace=True (default) returns None
+                    if inplace=False returns a new ContinuousFactor instance.
+
+        Examples
+        --------
+        >>> import numpy as np
+        >>> from scipy.special import beta
+        >>> from pgmpy.factors.distributions import CustomDistribution
+        >>> def custom_pdf(x, y, z):
+        ...     return z*(np.power(x, 1) * np.power(y, 2)) / beta(x, y)
+        >>> custom_dist = CustomDistribution(['x', 'y', 'z'], custom_pdf)
+        >>> custom_dist.variables
+        ['x', 'y', 'z']
+        >>> custom_dist.assignment(1, 2, 3)
+        24.0
+
+        >>> custom_dist.reduce([('y', 2)])
+        >>> custom_dist.variables
+        ['x', 'z']
+        >>> custom_dist.assignment(1, 3)
+        24.0
+        """
+        if not isinstance(values, (list, tuple, np.ndarray)):
+            raise TypeError("variables: Expected type: iterable, "
+                            "got: {var_type}".format(var_type=type(values)))
+
+        for var, value in values:
+            if var not in self.variables:
+                raise ValueError("{var} not in scope.".format(var=var))
+
+        phi = self if inplace else self.copy()
+
+        var_to_remove = [var for var, value in values]
+        var_to_keep = [var for var in self.variables if var not in var_to_remove]
+
+        reduced_var_index = [(self.variables.index(var), value) for var, value in values]
+        pdf = self.pdf
+
+        def reduced_pdf(*args, **kwargs):
```

Comment:
Move all the functions defined inside the methods as methods of the class. Also make them private(start the method name with `_`). This will help us in speeding up inference algorithms because methods like `reduce` would be called a lot of times and if we have these functions inside, these will be initialized everytime the method is called.